[Inductor] remove extra buffer creation in realize into

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110787

**Summary**
Fix the llama performance regression reported in https://github.com/pytorch/pytorch/issues/109874

**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_input_mutation5
python -u -m pytest -s -v test_torchinductor.py -k  test_zero_element_mutation
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
[Inductor] remove extra buffer creation in realize into

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] remove extra buffer creation in realize into"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] remove extra buffer creation in realize into"


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_input_mutation5
python -u -m pytest -s -v test_torchinductor.py -k  test_zero_element_mutation
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] remove extra buffer creation in realize into"


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_input_mutation5
python -u -m pytest -s -v test_torchinductor.py -k  test_zero_element_mutation
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -2337,23 +2337,32 @@ def realize_into(cls, src, dst):
+            # 1. If src is_user_of dst, we suppose it has been realize in previous
+            #    step of: V.graph.mark_buffer_mutated(dst.get_name())
+            # 2. The src must be FlexibleLayout, since it will be changed to MutationLayout with dst
+            need_copy = False
```
It seems to me like there are two possibilities. Either
1. `src` has another user and so we must copy to preserve correctness
2. `src` has no users other than the copy and so we would expect it to be eliminated by fusion

Do you know which case is happening in llama and if the second, why isn't this buffer being removed?

leslie-fang-intel replied:
- For llama, Compare the generated code before and after the regression, we find after regression there are 16 extra memory copy with buffers below from `buf288` to `buf333`

<img width="797" alt="image" src="https://github.com/pytorch/pytorch/assets/53841472/0c097947-d4a3-4cfd-b856-1dbc5e12506f">


- Take `buf333` and `arg91_1` as example:
  - Before regression, there is no `buf333`. The result will directly write into  `arg91_1` .
  - After regression, the result will first write into `buf333`, then copy into `arg91_1`  which is a mutation of another buffer `buf334`.
 
<img width="782" alt="image" src="https://github.com/pytorch/pytorch/assets/53841472/0f176029-adf7-4615-a094-9e03e4347c46">

leslie-fang-intel replied:
As for why `buf333` doesn't eliminate in `dead_node_elimination` https://github.com/pytorch/pytorch/blob/3ec33957eb8d774946d79db9d8d051be2c2c88cf/torch/_inductor/scheduler.py#L1285

- `buf333` has one user node of `buf334`.
```
buf333: SchedulerNode(ComputedBuffer)
buf333.writes = [MemoryDep('buf333', c0, {c0: 16777216})]
buf333.unmet_dependencies = [MemoryDep('buf264', c1 - 512, {c0: 32, c1: 524288})]
buf333.met_dependencies = 
    [   MemoryDep('arg91_1', c0, {c0: 16777216}),
        MemoryDep('arg91_1', c1, {c0: 32, c1: 524288})]
buf333.users = [NodeUser(node=SchedulerNode(name='buf334'), can_inplace=True, is_weak=False)]
buf333.group.device = cpu
buf333.group.iteration = ((32, 1024, 512), ())
buf333.sizes = ([32, 1024, 512], [])
class buf333_loop_body:
....
```
- `buf334` has user node of `output`.
```
buf334: SchedulerNode(ComputedBuffer)
buf334.writes = [MemoryDep('buf334', c0, {c0: 16777216})]
buf334.unmet_dependencies = [MemoryDep('buf333', c0, {c0: 16777216}), WeakDep(name='buf267')]
buf334.met_dependencies = [StarDep(name='arg91_1')]
buf334.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
buf334.group.device = cpu
buf334.group.iteration = ((16777216,), ())
buf334.sizes = ([16777216], [])
buf334.mutations = ['arg91_1']
class buf334_loop_body:
    var_ranges = {z0: 16777216}
    index0 = z0
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('buf333', get_index)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf334', get_index_1, load, None)
        return store
```

leslie-fang-intel replied:
The reason why `buf333` and `buf334` don't be fused is `var_ranges` of `buf333` is `(32, 1024, 512)`, but `var_ranges` of `buf334` is `(16777216,)` it fails the `can_fuse` check here: https://github.com/pytorch/pytorch/blob/ddb0c265110cfd053009dee34cbc5151bed8cddc/torch/_inductor/codegen/cpp.py#L2802

- Here is `buf333`
```
buf333: SchedulerNode(ComputedBuffer)
buf333.writes = [MemoryDep('buf333', c0, {c0: 16777216})]
buf333.unmet_dependencies = [MemoryDep('buf264', c1 - 512, {c0: 32, c1: 524288})]
buf333.met_dependencies = 
    [   MemoryDep('arg91_1', c0, {c0: 16777216}),
        MemoryDep('arg91_1', c1, {c0: 32, c1: 524288})]
buf333.users = [NodeUser(node=SchedulerNode(name='buf334'), can_inplace=True, is_weak=False)]
buf333.group.device = cpu
buf333.group.iteration = ((32, 1024, 512), ())
buf333.sizes = ([32, 1024, 512], [])
class buf333_loop_body:
    var_ranges = {z0: 32, z1: 1024, z2: 512}
   ....
```

- Here is `buf334`
```
buf334: SchedulerNode(ComputedBuffer)
buf334.writes = [MemoryDep('buf334', c0, {c0: 16777216})]
buf334.unmet_dependencies = [MemoryDep('buf333', c0, {c0: 16777216}), WeakDep(name='buf267')]
buf334.met_dependencies = [StarDep(name='arg91_1')]
buf334.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
buf334.group.device = cpu
buf334.group.iteration = ((16777216,), ())
buf334.sizes = ([16777216], [])
buf334.mutations = ['arg91_1']
class buf334_loop_body:
    var_ranges = {z0: 16777216}
    ....
```

peterbell10 replied:
In that case I think there are two ways forward, we either need to:
1. Prove that the buffer has no users and skip the copy
2. Enable cpu inductor to make this type of fusion

I think 1 should be possible by adding a special IR node for `Copy` and adding a copy-removal pass in the scheduler. Perhaps you could comment on 2?

leslie-fang-intel replied:
Thanks for the comment. Some questions I am not understanding clearly:

- _Actually in this case the buffer also has other users_ 
Looks like here `buf333` has only one user of `buf334`. And `buf334` has only one user of `Output`.  What's the `other users` discussed here?

- _aliasing is the only way to completely remove the redundant stores_
Did you mean making `buf334` of `buf333`'s alias like `buf334.data.layout = MutationLayout(buf333)`

jansel reviewed (changes requested):
tests are failing

leslie-fang-intel committed:
Update on "[Inductor] remove extra buffer creation in realize into"


**Summary**
Fix the llama performance regression reported in https://github.com/pytorch/pytorch/issues/109874

**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_input_mutation5
python -u -m pytest -s -v test_torchinductor.py -k  test_zero_element_mutation
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -2332,28 +2332,55 @@ def realize_into(cls, src, dst):
+            and isinstance(src.get_layout(), FlexibleLayout)
+            and not src.is_zero_elements()
+            and (
+                (src.get_name() in realized_bufs)
```
If `src` was realized doesn't that imply `src` is a user of `dst` and so copying is necessary in case the indexing doesn't line up. For example something like `x.copy_(x.transpose())` is unlikely to work.

leslie-fang-intel replied:
Thanks for the comments.

- `x.copy_(x.transpose())` may not be a good example. Seems it will also fail in eager mode. And 2 tensors with different sizes may not be copied.

- Another way comes to my mind may be as the complement of https://github.com/pytorch/pytorch/pull/107614
If src will be realized here, https://github.com/pytorch/pytorch/blob/e942fddb83c0a101e0bb10fd66538cf58254bd5f/torch/_inductor/ir.py#L2368
Can we skip the src realization in https://github.com/pytorch/pytorch/blob/e942fddb83c0a101e0bb10fd66538cf58254bd5f/torch/_inductor/ir.py#L2347
I tried this idea, it can also fix the regression in https://github.com/pytorch/pytorch/issues/109874

peterbell10 replied:
> `x.copy_(x.transpose())` may not be a good example. Seems it will also fail in eager mode.

Okay, the same principle applies to `x.transpose().clone()`.

>  Can we skip the src realization in [mark_buffer_mutated]

No I don't think this is sound. If `src` is a user of `dst` and `src` isn't realized before the mutation then other users of `src` may end up doing the calculation with the post-mutation version of `dst`.

leslie-fang-intel committed:
Update on "[Inductor] remove extra buffer creation in realize into"


**Summary**
Fix the llama performance regression reported in https://github.com/pytorch/pytorch/issues/109874

**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_input_mutation5
python -u -m pytest -s -v test_torchinductor.py -k  test_zero_element_mutation
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
